core/trino-main/src/main/java/io/trino/operator/StreamingDirectExchangeBuffer.java

```diff
@@ -69,9 +70,15 @@ public StreamingDirectExchangeBuffer(Executor executor, DataSize bufferCapacity)
     }
 
     @Override
-    public ListenableFuture<Void> isBlocked()
+    public synchronized ListenableFuture<Void> isBlocked()
```

Comment:
There is a comment in `ExchangeOperator::isBlocked`:
```
// Avoid registering a new callback in the data source when one is already pending
```

> It would be better for caller to identify itself via isBlocked() argument. 

this was a first approach